[2025-04-26T02:57:56.543+0000] {taskinstance.py:1956} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: traffic_data_pipeline.submit_traffic_processor_spark_job manual__2025-04-26T02:57:54.641103+00:00 [queued]>
[2025-04-26T02:57:56.562+0000] {taskinstance.py:1956} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: traffic_data_pipeline.submit_traffic_processor_spark_job manual__2025-04-26T02:57:54.641103+00:00 [queued]>
[2025-04-26T02:57:56.563+0000] {taskinstance.py:2170} INFO - Starting attempt 1 of 1
[2025-04-26T02:57:56.588+0000] {taskinstance.py:2191} INFO - Executing <Task(BashOperator): submit_traffic_processor_spark_job> on 2025-04-26 02:57:54.641103+00:00
[2025-04-26T02:57:56.606+0000] {standard_task_runner.py:60} INFO - Started process 504 to run task
[2025-04-26T02:57:56.611+0000] {standard_task_runner.py:87} INFO - Running: ['***', 'tasks', 'run', 'traffic_data_pipeline', 'submit_traffic_processor_spark_job', 'manual__2025-04-26T02:57:54.641103+00:00', '--job-id', '6', '--raw', '--subdir', 'DAGS_FOLDER/traffic_pipeline_dag.py', '--cfg-path', '/tmp/tmpzrix9b44']
[2025-04-26T02:57:56.615+0000] {standard_task_runner.py:88} INFO - Job 6: Subtask submit_traffic_processor_spark_job
[2025-04-26T02:57:56.735+0000] {task_command.py:423} INFO - Running <TaskInstance: traffic_data_pipeline.submit_traffic_processor_spark_job manual__2025-04-26T02:57:54.641103+00:00 [running]> on host f81ac7673fa0
[2025-04-26T02:57:57.238+0000] {taskinstance.py:2480} INFO - Exporting env vars: AIRFLOW_CTX_DAG_OWNER='***' AIRFLOW_CTX_DAG_ID='traffic_data_pipeline' AIRFLOW_CTX_TASK_ID='submit_traffic_processor_spark_job' AIRFLOW_CTX_EXECUTION_DATE='2025-04-26T02:57:54.641103+00:00' AIRFLOW_CTX_TRY_NUMBER='1' AIRFLOW_CTX_DAG_RUN_ID='manual__2025-04-26T02:57:54.641103+00:00'
[2025-04-26T02:57:57.241+0000] {subprocess.py:63} INFO - Tmp dir root location: /tmp
[2025-04-26T02:57:57.243+0000] {subprocess.py:75} INFO - Running command: ['/usr/bin/bash', '-c', '\n            export KAFKA_BROKERS_INTERNAL="kafk-1:9092,kafk-2:9092,kafk-3:9092" && \\\n            export KAFKA_TOPIC="raw_traffic_data" && \\\n            export SCHEMA_REGISTRY_URL="http://schema-registry:8081" && \\\n            export REDIS_HOST="redis" && \\\n            export CLICKHOUSE_HOST="clickhouse-server" && \\\n            export CLICKHOUSE_NATIVE_PORT="9000" && \\\n            export MINIO_ENDPOINT="http://minio:9000" && \\\n            export MINIO_ACCESS_KEY="minioadmin" && \\\n            export MINIO_SECRET_KEY="minioadmin" && \\\n            export MINIO_BUCKET="traffic-data" && \\\n            export CHECKPOINT_LOCATION="s3a://traffic-data/checkpoints_***" && \\\n            export PYSPARK_PYTHON=/usr/bin/python3 && \\\n            spark-submit \\\n            --master local[*] \\\n            --deploy-mode client \\\n            --conf spark.sql.streaming.schemaInference=true \\\n            --conf spark.hadoop.fs.s3a.endpoint=http://minio:9000 \\\n            --conf spark.hadoop.fs.s3a.access.key=minioadmin \\\n            --conf spark.hadoop.fs.s3a.secret.key=minioadmin \\\n            --conf spark.hadoop.fs.s3a.path.style.access=true \\\n            --conf spark.hadoop.fs.s3a.impl=org.apache.hadoop.fs.s3a.S3AFileSystem \\\n            --conf spark.sql.avro.schemaRegistryUrl=http://schema-registry:8081 \\\n            /opt/spark/app/processor.py\n            ']
[2025-04-26T02:57:57.255+0000] {subprocess.py:86} INFO - Output:
[2025-04-26T02:57:57.269+0000] {subprocess.py:93} INFO - /opt/spark/bin/load-spark-env.sh: line 68: ps: command not found
[2025-04-26T02:58:01.272+0000] {subprocess.py:93} INFO - --- RUNNING LATEST processor.py with IMPORTED haversine ---
[2025-04-26T02:58:01.423+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 INFO SparkContext: Running Spark version 3.3.4
[2025-04-26T02:58:01.528+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
[2025-04-26T02:58:01.652+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 INFO ResourceUtils: ==============================================================
[2025-04-26T02:58:01.653+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 INFO ResourceUtils: No custom resources configured for spark.driver.
[2025-04-26T02:58:01.654+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 INFO ResourceUtils: ==============================================================
[2025-04-26T02:58:01.655+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 INFO SparkContext: Submitted application: TrafficProcessor
[2025-04-26T02:58:01.683+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 INFO ResourceProfile: Default ResourceProfile created, executor resources: Map(cores -> name: cores, amount: 1, script: , vendor: , memory -> name: memory, amount: 1024, script: , vendor: , offHeap -> name: offHeap, amount: 0, script: , vendor: ), task resources: Map(cpus -> name: cpus, amount: 1.0)
[2025-04-26T02:58:01.701+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 INFO ResourceProfile: Limiting resource is cpu
[2025-04-26T02:58:01.702+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 INFO ResourceProfileManager: Added ResourceProfile id: 0
[2025-04-26T02:58:01.775+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 INFO SecurityManager: Changing view acls to: ***
[2025-04-26T02:58:01.776+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 INFO SecurityManager: Changing modify acls to: ***
[2025-04-26T02:58:01.777+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 INFO SecurityManager: Changing view acls groups to:
[2025-04-26T02:58:01.778+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 INFO SecurityManager: Changing modify acls groups to:
[2025-04-26T02:58:01.779+0000] {subprocess.py:93} INFO - 25/04/26 02:58:01 INFO SecurityManager: SecurityManager: authentication disabled; ui acls disabled; users  with view permissions: Set(***); groups with view permissions: Set(); users  with modify permissions: Set(***); groups with modify permissions: Set()
[2025-04-26T02:58:02.037+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO Utils: Successfully started service 'sparkDriver' on port 40503.
[2025-04-26T02:58:02.071+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO SparkEnv: Registering MapOutputTracker
[2025-04-26T02:58:02.126+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO SparkEnv: Registering BlockManagerMaster
[2025-04-26T02:58:02.148+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO BlockManagerMasterEndpoint: Using org.apache.spark.storage.DefaultTopologyMapper for getting topology information
[2025-04-26T02:58:02.149+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO BlockManagerMasterEndpoint: BlockManagerMasterEndpoint up
[2025-04-26T02:58:02.154+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO SparkEnv: Registering BlockManagerMasterHeartbeat
[2025-04-26T02:58:02.179+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO DiskBlockManager: Created local directory at /tmp/blockmgr-d6a0c4d7-c001-4e5f-b06e-1308b2dad9d2
[2025-04-26T02:58:02.221+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO MemoryStore: MemoryStore started with capacity 434.4 MiB
[2025-04-26T02:58:02.256+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO SparkEnv: Registering OutputCommitCoordinator
[2025-04-26T02:58:02.466+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO Utils: Successfully started service 'SparkUI' on port 4040.
[2025-04-26T02:58:02.565+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO Executor: Starting executor ID driver on host f81ac7673fa0
[2025-04-26T02:58:02.572+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO Executor: Starting executor with user classpath (userClassPathFirst = false): ''
[2025-04-26T02:58:02.592+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO Utils: Successfully started service 'org.apache.spark.network.netty.NettyBlockTransferService' on port 40495.
[2025-04-26T02:58:02.593+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO NettyBlockTransferService: Server created on f81ac7673fa0:40495
[2025-04-26T02:58:02.594+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO BlockManager: Using org.apache.spark.storage.RandomBlockReplicationPolicy for block replication policy
[2025-04-26T02:58:02.600+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO BlockManagerMaster: Registering BlockManager BlockManagerId(driver, f81ac7673fa0, 40495, None)
[2025-04-26T02:58:02.605+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO BlockManagerMasterEndpoint: Registering block manager f81ac7673fa0:40495 with 434.4 MiB RAM, BlockManagerId(driver, f81ac7673fa0, 40495, None)
[2025-04-26T02:58:02.608+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO BlockManagerMaster: Registered BlockManager BlockManagerId(driver, f81ac7673fa0, 40495, None)
[2025-04-26T02:58:02.609+0000] {subprocess.py:93} INFO - 25/04/26 02:58:02 INFO BlockManager: Initialized BlockManager: BlockManagerId(driver, f81ac7673fa0, 40495, None)
[2025-04-26T02:58:03.042+0000] {subprocess.py:93} INFO - Spark Session created. Version: 3.3.4
[2025-04-26T02:58:03.043+0000] {subprocess.py:93} INFO - Broadcasting segments data...
[2025-04-26T02:58:03.228+0000] {subprocess.py:93} INFO - Segments data broadcasted.
[2025-04-26T02:58:03.229+0000] {subprocess.py:93} INFO - Reading from Kafka: kafk-1:9092,kafk-2:9092,kafk-3:9092 / Topic: raw_traffic_data
[2025-04-26T02:58:03.230+0000] {subprocess.py:93} INFO - Kafka options: {'kafka.bootstrap.servers': 'kafk-1:9092,kafk-2:9092,kafk-3:9092', 'subscribe': 'raw_traffic_data', 'startingOffsets': 'latest', 'failOnDataLoss': 'false'}
[2025-04-26T02:58:06.144+0000] {subprocess.py:93} INFO - Reading Avro schema from file: /opt/spark/schemas/raw_traffic_event.avsc
[2025-04-26T02:58:06.148+0000] {subprocess.py:93} INFO - Successfully read Avro schema string.
[2025-04-26T02:58:06.149+0000] {subprocess.py:93} INFO - Deserializing Kafka messages using schema string from file...
[2025-04-26T02:58:06.429+0000] {subprocess.py:93} INFO - Traceback (most recent call last):
[2025-04-26T02:58:06.430+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/serializers.py", line 458, in dumps
[2025-04-26T02:58:06.431+0000] {subprocess.py:93} INFO -     return cloudpickle.dumps(obj, pickle_protocol)
[2025-04-26T02:58:06.432+0000] {subprocess.py:93} INFO -            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2025-04-26T02:58:06.432+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle_fast.py", line 73, in dumps
[2025-04-26T02:58:06.433+0000] {subprocess.py:93} INFO -     cp.dump(obj)
[2025-04-26T02:58:06.434+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle_fast.py", line 602, in dump
[2025-04-26T02:58:06.434+0000] {subprocess.py:93} INFO -     return Pickler.dump(self, obj)
[2025-04-26T02:58:06.435+0000] {subprocess.py:93} INFO -            ^^^^^^^^^^^^^^^^^^^^^^^
[2025-04-26T02:58:06.435+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle_fast.py", line 692, in reducer_override
[2025-04-26T02:58:06.436+0000] {subprocess.py:93} INFO -     return self._function_reduce(obj)
[2025-04-26T02:58:06.436+0000] {subprocess.py:93} INFO -            ^^^^^^^^^^^^^^^^^^^^^^^^^^
[2025-04-26T02:58:06.437+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle_fast.py", line 565, in _function_reduce
[2025-04-26T02:58:06.438+0000] {subprocess.py:93} INFO -     return self._dynamic_function_reduce(obj)
[2025-04-26T02:58:06.438+0000] {subprocess.py:93} INFO -            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2025-04-26T02:58:06.439+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle_fast.py", line 546, in _dynamic_function_reduce
[2025-04-26T02:58:06.440+0000] {subprocess.py:93} INFO -     state = _function_getstate(func)
[2025-04-26T02:58:06.441+0000] {subprocess.py:93} INFO -             ^^^^^^^^^^^^^^^^^^^^^^^^
[2025-04-26T02:58:06.441+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle_fast.py", line 157, in _function_getstate
[2025-04-26T02:58:06.442+0000] {subprocess.py:93} INFO -     f_globals_ref = _extract_code_globals(func.__code__)
[2025-04-26T02:58:06.443+0000] {subprocess.py:93} INFO -                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2025-04-26T02:58:06.444+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle.py", line 334, in _extract_code_globals
[2025-04-26T02:58:06.444+0000] {subprocess.py:93} INFO -     out_names = {names[oparg]: None for _, oparg in _walk_global_ops(co)}
[2025-04-26T02:58:06.445+0000] {subprocess.py:93} INFO -                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2025-04-26T02:58:06.446+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle.py", line 334, in <dictcomp>
[2025-04-26T02:58:06.446+0000] {subprocess.py:93} INFO -     out_names = {names[oparg]: None for _, oparg in _walk_global_ops(co)}
[2025-04-26T02:58:06.447+0000] {subprocess.py:93} INFO -                  ~~~~~^^^^^^^
[2025-04-26T02:58:06.447+0000] {subprocess.py:93} INFO - IndexError: tuple index out of range
[2025-04-26T02:58:06.448+0000] {subprocess.py:93} INFO - Traceback (most recent call last):
[2025-04-26T02:58:06.449+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/serializers.py", line 458, in dumps
[2025-04-26T02:58:06.450+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle_fast.py", line 73, in dumps
[2025-04-26T02:58:06.450+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle_fast.py", line 602, in dump
[2025-04-26T02:58:06.451+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle_fast.py", line 692, in reducer_override
[2025-04-26T02:58:06.452+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle_fast.py", line 565, in _function_reduce
[2025-04-26T02:58:06.453+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle_fast.py", line 546, in _dynamic_function_reduce
[2025-04-26T02:58:06.454+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle_fast.py", line 157, in _function_getstate
[2025-04-26T02:58:06.454+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle.py", line 334, in _extract_code_globals
[2025-04-26T02:58:06.455+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle/cloudpickle.py", line 334, in <dictcomp>
[2025-04-26T02:58:06.456+0000] {subprocess.py:93} INFO - IndexError: tuple index out of range
[2025-04-26T02:58:06.456+0000] {subprocess.py:93} INFO - 
[2025-04-26T02:58:06.457+0000] {subprocess.py:93} INFO - During handling of the above exception, another exception occurred:
[2025-04-26T02:58:06.457+0000] {subprocess.py:93} INFO - 
[2025-04-26T02:58:06.459+0000] {subprocess.py:93} INFO - Traceback (most recent call last):
[2025-04-26T02:58:06.459+0000] {subprocess.py:93} INFO -   File "/opt/spark/app/processor.py", line 263, in <module>
[2025-04-26T02:58:06.460+0000] {subprocess.py:93} INFO -     main()
[2025-04-26T02:58:06.460+0000] {subprocess.py:93} INFO -   File "/opt/spark/app/processor.py", line 219, in main
[2025-04-26T02:58:06.461+0000] {subprocess.py:93} INFO -     find_nearest_segment_udf_new(col("latitude"), col("longitude"))
[2025-04-26T02:58:06.462+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/sql/udf.py", line 276, in wrapper
[2025-04-26T02:58:06.462+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/sql/udf.py", line 249, in __call__
[2025-04-26T02:58:06.463+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/sql/udf.py", line 215, in _judf
[2025-04-26T02:58:06.463+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/sql/udf.py", line 224, in _create_judf
[2025-04-26T02:58:06.464+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/sql/udf.py", line 50, in _wrap_function
[2025-04-26T02:58:06.464+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/rdd.py", line 3345, in _prepare_for_python_RDD
[2025-04-26T02:58:06.465+0000] {subprocess.py:93} INFO -   File "/opt/spark/python/lib/pyspark.zip/pyspark/serializers.py", line 468, in dumps
[2025-04-26T02:58:06.465+0000] {subprocess.py:93} INFO - _pickle.PicklingError: Could not serialize object: IndexError: tuple index out of range
[2025-04-26T02:58:06.621+0000] {subprocess.py:97} INFO - Command exited with return code 1
[2025-04-26T02:58:06.644+0000] {taskinstance.py:2698} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/models/taskinstance.py", line 433, in _execute_task
    result = execute_callable(context=context, **execute_callable_kwargs)
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/operators/bash.py", line 212, in execute
    raise AirflowException(
airflow.exceptions.AirflowException: Bash command failed. The command returned a non-zero exit code 1.
[2025-04-26T02:58:06.649+0000] {taskinstance.py:1138} INFO - Marking task as FAILED. dag_id=traffic_data_pipeline, task_id=submit_traffic_processor_spark_job, execution_date=20250426T025754, start_date=20250426T025756, end_date=20250426T025806
[2025-04-26T02:58:06.673+0000] {standard_task_runner.py:107} ERROR - Failed to execute job 6 for task submit_traffic_processor_spark_job (Bash command failed. The command returned a non-zero exit code 1.; 504)
[2025-04-26T02:58:06.688+0000] {local_task_job_runner.py:234} INFO - Task exited with return code 1
[2025-04-26T02:58:06.724+0000] {taskinstance.py:3280} INFO - 0 downstream tasks scheduled from follow-on schedule check
